make/mt-config/tests/mt-gtest.mk

```diff
@@ -57,7 +57,12 @@ GTEST_OBJS := $(GTEST_OBJDIR)/gtest-all.o \
 # Set up iotc googletest tests.
 IOTC_GTEST_SUITE ?= iotc_gtests
 IOTC_GTEST_OBJDIR := $(IOTC_TEST_OBJDIR)/gtests
-IOTC_GTEST_SOURCES = $(shell find $(LIBIOTC)/src -name "*.cc")
+
+# Test cases: files ending in _test.cc or _test_something.cc
+IOTC_GTEST_SOURCES = $(shell find $(LIBIOTC)/src -regex ".+_test.cc" -o -regex ".+_test_\w+.cc")
+
+# Test utils: .cc files in src/tests/
+IOTC_GTEST_SOURCES += $(shell find $(LIBIOTC)/src/tests -name "*.cc")
```

Comment:
There is a more gnu make standard way ([wildcard example in this repo](https://github.com/GoogleCloudPlatform/iot-device-sdk-embedded-c/blob/master/make/mt-config/mt-config.mk#L229)) to collect all .cc files and a standard gnu make command: [`filter`](https://www.gnu.org/software/make/manual/html_node/Text-Functions.html) to get only test files.

Reply:
Also, would be good to update the [collection of non-test .cc files](https://github.com/GoogleCloudPlatform/iot-device-sdk-embedded-c/blob/master/make/mt-config/mt-config.mk#L229).
Or even better to have a single place where all .cc files are collected and filtered into two disjunct set of files: test and non-test files. All this to ensure they are always in sync.

Reply:
> There is a more gnu make standard way ([wildcard example in this repo](https://github.com/GoogleCloudPlatform/iot-device-sdk-embedded-c/blob/master/make/mt-config/mt-config.mk#L229)) to collect all .cc files and a standard gnu make command: [`filter`](https://www.gnu.org/software/make/manual/html_node/Text-Functions.html) to get only test files.

Good suggestion. Although this find command should be fairly portable.

I tried to re-implement using `wildcard`, but that would require change the first line with `find` into 4 lines:
```
IOTC_GTEST_SOURCES := $(wildcard ./src/*_test.cc)
IOTC_GTEST_SOURCES += $(wildcard ./src/*_test_*.cc)
IOTC_GTEST_SOURCES += $(foreach layerdir,$(IOTC_SRCDIRS), $(wildcard $(layerdir)/*_test.cc))
IOTC_GTEST_SOURCES += $(foreach layerdir,$(IOTC_SRCDIRS), $(wildcard $(layerdir)/*_test_*.cc))
```

or did I miss something ?

Reply:
> Also, would be good to update the [collection of non-test .cc files](https://github.com/GoogleCloudPlatform/iot-device-sdk-embedded-c/blob/master/make/mt-config/mt-config.mk#L229).
> Or even better to have a single place where all .cc files are collected and filtered into two disjunct set of files: test and non-test files. All this to ensure they are always in sync.

+1
By any chance do you also have an idea how to achieve that?
I was thinking to define `IOTC_SOURCES_CXX` as `{all .cc} - IOTC_GTEST_SOURCES` 
(do something like `IOTC_SOURCES_CXX := $(filter-out $(IOTC_GTEST_SOURCES) $(IOTC_SOURCES_CXX))`)
But that does not work, `filter-out` first argument is a string. Any idea how to compute set difference with Makefile ?

Reply:
> > Also, would be good to update the [collection of non-test .cc files](https://github.com/GoogleCloudPlatform/iot-device-sdk-embedded-c/blob/master/make/mt-config/mt-config.mk#L229).
> > Or even better to have a single place where all .cc files are collected and filtered into two disjunct set of files: test and non-test files. All this to ensure they are always in sync.
> 
> +1
> By any chance do you also have an idea how to achieve that?
> I was thinking to define `IOTC_SOURCES_CXX` as `{all .cc} - IOTC_GTEST_SOURCES`
> (do something like `IOTC_SOURCES_CXX := $(filter-out $(IOTC_GTEST_SOURCES) $(IOTC_SOURCES_CXX))`)
> But that does not work, `filter-out` first argument is a string. Any idea how to compute set difference with Makefile ?

Never mind. Please unread what I wrote above.

Reply:
I think the first two lines:
```
IOTC_GTEST_SOURCES := $(wildcard ./src/*_test.cc)
IOTC_GTEST_SOURCES += $(wildcard ./src/*_test_*.cc)
``` don't do much since the `./src` directory is empty. So they can be removed from the solution.

So, although I could accept the 'shell find' since nothing wrong with it, I'd rather apply a simplification (strict unification of test filenames) and still stick to built in gnu make functions:
1) for simplicity I'd rename the *_test_XXX.cc tests to *_XXX_test.cc
2) then a single line would do the rest:  `IOTC_GTEST_SOURCES += $(foreach srcdir,$(IOTC_SRCDIRS), $(wildcard $(srcdir)/*_test.cc))`

or without file rename:
`IOTC_GTEST_SOURCES += $(foreach srcdir,$(IOTC_SRCDIRS), $(wildcard $(srcdir)/*_test*.cc))`

Again I can accept the `shell find` solution if you prefer since running a shell in make on linux machine doesn't blow up the execution time. Just the reasoning is what I cannot accept: there is no need for 4 lines with built-in function. The "inconsistent" test file naming and using empty dir caused the 1 line to become 4 lines. And I'd use `shell` command for things which don't have their built-in pairs.

Reply:
Please also put all of the tests in their own separate tests directory.  We spoke of this in the past but yet we're continuing to add tests to the main source directories.  While the hosting of test sources alongside the library sources is a google best practice, it adds a extra housekeeping step for those who want to copy /import only the library sources into their IDE.

This is not just personal opinion (as it was when we first discussed this.)  We've also received partner feedback about this.

Reply:
You clearly pointed out the root of the problem: the naming inconsistency.
Renamed so that all test files end with the `_test.cc` prefix.

Reply:
Upon further inspection, we have test sources that are more than 1 sub-directory deep.

GNU Make does have a recursive callback, but this is not portable (does not work on Mac OS for example). I suggest keeping the `find` command as the most concise yet portable solution.
And we might revert this whole structure anyway.

Reply:
ok